src/ClimaLSM.jl

```diff
@@ -1,5 +1,164 @@
 module ClimaLSM
 
-greet() = print("Hello World!")
+using ClimaCore
+import ClimaCore: Fields
+include("Domains.jl")
+using .Domains
+
+export AbstractModel,
+    make_rhs,
+    make_ode_function,
+    make_update_aux,
+    initialize_prognostic,
+    initialize_auxiliary,
+    initialize_states,
+    LandModel,
+    prognostic_vars,
+    auxiliary_vars
+## Default methods for all models - to be in a seperate module at some point.
+
+abstract type AbstractModel{FT <: AbstractFloat} end
+
+"""
+   prognostic_vars(m::AbstractModel)
+
+Returns the prognostic variable symbols for the model in the form of a tuple.
+"""
+prognostic_vars(m::AbstractModel) = ()
+
+"""
+   auxiliary_vars(m::AbstractModel)
+
+Returns the auxiliary variable symbols for the model in the form of a tuple.
+"""
+auxiliary_vars(m::AbstractModel) = ()
+
+"""
+    make_rhs(model::AbstractModel)
+
+Return a `rhs!` function that updates state variables.
+
+`rhs!` should be compatible with OrdinaryDiffEq.jl solvers.
+"""
+function make_rhs(model::AbstractModel)
+    function rhs!(dY, Y, p, t) end
+    return rhs!
+end
+
+"""
+    make_update_aux(model::AbstractModel)
+
+Return an `update_aux!` function that updates auxiliary parameters `p`.
+"""
+function make_update_aux(model::AbstractModel)
+    function update_aux!(p, Y, t) end
+    return update_aux!
+end
+
+"""
+    make_ode_function(model::AbstractModel)
+
+Returns an `ode_function` that updates auxiliary variables and
+updates the prognostic state.
+
+`ode_function!` should be compatible with OrdinaryDiffEq.jl solvers.
+"""
+function make_ode_function(model::AbstractModel)
+    rhs! = make_rhs(model)
+    update_aux! = make_update_aux(model)
+    function ode_function!(dY, Y, p, t)
+        update_aux!(p, Y, t)
+        rhs!(dY, Y, p, t)
+    end
+    return ode_function!
+end
+
+"""
+    initialize_prognostic(model::AbstractModel,_)
+
+Returns a FieldVector of prognostic variables for `model` with the required
+structure.  This is a stub. We will change it, if possible,
+ as we determine what components require.
+An idea is to use something like `similar(coords)`.
+"""
+function initialize_prognostic(model::AbstractModel, _)
```

Comment:
@jb-mackay in the end, it may be best for each submodel to define this method and not have all submodels be using the same default - but we can revisit. It actually might be nice to have the fallback/default be not adding *any* variables, because some submodels may have no aux vars, or even no prognostic vars (if a certain submodel involves only algebraic equations, e.g., not a differential equation).

Reply:
Makes sense to me, I think we should definitely start there and then see if we can generalize once we have a few examples.

Reply:
see https://github.com/CliMA/ClimaLSM.jl/issues/3, will leave as is.